confidential-data-hub/hub/src/bin/confidential-data-hub/config.rs

```diff
@@ -0,0 +1,138 @@
+
+#[derive(Deserialize, Debug)]
+pub struct CdhConfig {
+    pub kbc: KbsConfig,
+
+    pub credentials: Vec<Credential>,
+
+    pub socket: String,
+}
+
+impl Default for CdhConfig {
+    fn default() -> Self {
+        Self {
+            socket: DEFAULT_CDH_SOCKET_ADDR.into(),
+            kbc: KbsConfig::default(),
+            credentials: Vec::default(),
+        }
+    }
+}
+
+impl CdhConfig {
+    pub async fn init(config_path: &str) -> Result<Self> {
+        match Self::from_file(config_path) {
+            std::result::Result::Ok(mut config) => {
+                info!("Read config file from {config_path}");
+                config.update_from_kernel_cmdline().await?;
+                return Ok(config);
+            }
+            Err(e) => warn!("read config file {config_path} failed {e:?}"),
+        }
+
+        warn!("Use default config. aa_kbc_params = offline_fs_kbc::null");
+        let mut config = CdhConfig::default();
+        config.update_from_kernel_cmdline().await?;
+        Ok(config)
+    }
+
+    /// Load `CdhConfig` from a configuration file. Supported formats are all formats supported by the
+    /// `config` crate.
+    fn from_file(config_path: &str) -> Result<Self> {
+        let c = Config::builder()
+            .set_default("socket", DEFAULT_CDH_SOCKET_ADDR)?
+            .add_source(File::with_name(config_path))
+            .build()?;
+
+        let res = c.try_deserialize().context("invalid config").unwrap();
+        Ok(res)
+    }
+
+    /// all the resource ids can be from the kernel commandline in the following format:
+    /// ```shell
+    /// cdh.kbs_resources=<resource id 1>::<target path 1>,<resource id 2>::<target path 2>...
+    /// ```
+    ///
+    /// for example
+    /// ```shell
+    /// cdh.kbs_resources=kbs:///default/key/1::/run/temp1,kbs:///default/key/2::/run/temp2
+    /// ```
+    ///
+    /// It is supposed that all the `target path` should be with prefix
+    /// `/run/confidential-containers/cdh` or it will be treated as dangerous
+    /// path.
+    ///
+    /// TODO: delete this way after initdata mechanism could cover CDH's launch config.
+    async fn update_from_kernel_cmdline(&mut self) -> Result<()> {
+        let cmdline = fs::read_to_string("/proc/cmdline")
+            .await
+            .context("read kernel cmdline failed")?;
+        let kbs_resources = cmdline
+            .split_ascii_whitespace()
+            .find(|para| para.starts_with("cdh.kbs_resources="))
+            .unwrap_or("cdh.kbs_resources=")
+            .strip_prefix("cdh.kbs_resources=")
+            .expect("must have one")
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .filter_map(|it| it.split_once("::"))
+            .map(|it| Credential {
+                resource_uri: it.0.to_owned(),
+                path: it.1.to_owned(),
+            });
+
+        self.credentials.extend(kbs_resources);
+        Ok(())
+    }
+}
+
+impl CdhConfig {
+    pub fn apply(&self) {
```

Comment:
This is kind of a vague name. Is this function setting up the vars for the AA? Might be better to rename.

Reply:
I once want to specify more than one envs, including `KBS_URL`, `KBC`, `KBS_PUB_KEY` etc. `set_configuration_envs()` might be a better name?

Reply:
Let me temporarily use the name `set_configuration_envs()` in the fixup. Please give any suggestions if it is not perfect.